File: csharp/elons-toys/ElonsToys.cs
using System;

class RemoteControlCar
{
    private int _distanceTravelled = 0;
    private int _chargeRemaining = 100;

    public static RemoteControlCar Buy() => new RemoteControlCar();

    public string DistanceDisplay() => $"Driven {_distanceTravelled} meters";

    public string BatteryDisplay()
    {
        if (_chargeRemaining == 0) {
        return "Battery empty";
        }

        return $"Battery at {_chargeRemaining}%";
    }

    public void Drive()
    {
        if (_chargeRemaining == 0) {
            return;
        }

        _distanceTravelled = _distanceTravelled + 20;
        _chargeRemaining = _chargeRemaining - 1;
    }
}

File: go/elons-toys/elons_toys.go
package elon

import "fmt"

// Drive drains a Car's battery by batteryDrain and increases its distance
// driven by speed
func (car *Car) Drive() {
	if car.battery-car.batteryDrain <= 0 {
		return
	}

	car.distance += car.speed
	car.battery -= car.batteryDrain
}

// DisplayDistance displays a Car's distance driven
func (car *Car) DisplayDistance() string {
	return fmt.Sprintf("Driven %d meters", car.distance)
}

// DisplayBattery displays the current battery info of a Car
func (car *Car) DisplayBattery() string {
	return fmt.Sprintf("Battery at %d%%", car.battery)
}

// CanFinish returns a bool describing whether or not a car can finish a track
func (car *Car) CanFinish(trackDistance int) bool {
	unitsUntilDrained := car.battery / car.batteryDrain
	driveableDistance := car.speed * unitsUntilDrained

	return driveableDistance >= trackDistance
}
